### PR TITLE
[BEAM-3147] Nexmark launcher refactor

### PR DESCRIPTION
**Pull request by akedin**
Refactoring Nexmark launcher. Goal is to make it more modular and extensible to simplify [SQL integration](https://issues.apache.org/jira/browse/BEAM-3147). This PR mostly takes pieces out of NexmarkLauncher, like sinks or event sources, into their own PTransforms.

This version runs as expected in batch mode on direct runner. Working on testing other modes.

Currently this PR also includes commits from the [previous PR](https://github.com/apache/beam/pull/4087).

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by akedin** [comment]
@kennknowles @xumingming , please take a look

**Comment by xumingming** [comment]
Looks like you are using the same issue id: `BEAM-3147` for two PRs, it might be better to split this `BEAM-3147` issue into several smaller issue to better tracking things.

File: sdks/java/nexmark/src/test/java/org/apache/beam/sdk/nexmark/sources/EventSourceFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+/**
+ * Unit tests for {@link EventSourceFactoryTest}.
```
Comment (xumingming) [inline sdks/java/nexmark/src/test/java/org/apache/beam/sdk/nexmark/sources/EventSourceFactoryTest.java]: Unit tests for **{@link EventSourceFactory}**

**Comment by iemejia** [comment]
Can you please explain the motivation of such a big refactor, From a quick look I have the impression that the rewrite is more of a way to understand the base code than really needed for your intended goal of supporting the SQL. Of course there are some interesting pieces there but not sure if this huge refactor is really needed.

**Comment by akedin** [comment]
@iemejia , you're right, I was doing it while familiarizing myself with the code. And technically speaking it probably is not needed for the implementation of SQL. My reasoning was that it would be a good opportunity to address few issues while I'm at it.

But I feel it is necessary to refactor this at some point anyway. [This class](https://github.com/apache/beam/blob/64ff21f35ee2946d38645fb0a51678628e49e62a/sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/NexmarkLauncher.java) is 1157 lines of procedural code, with methods of 100+ lines long, with minimal logical structure. And it has zero test coverage from what I see. It is hard to follow, modify and debug.

If there's an opposition to this amount of changes at once, then I can drop this for now and try to add new features to existing code in smaller chunks.

**Comment by kennknowles** [comment]
Since there are many commits here, perhaps just a few can be reviewed at a time and then it will be digestible? I strongly agree with @akedin that this needs an overhaul to be maintainable.

**Comment by iemejia** [comment]
I agree that there is definitely space for maintainability improvements, that's why I asked for a more clear explanation for the goal of the changes. @akedin already did a good job by separating features per commit. So I think I understand some.
I agree with Kenn, maybe it is easier to divide the review in chunks, sorry if this fact + the discussion makes the review a bit longer than expected but please understand that 61 changed files / 6 KLOC of changes is a good amount to review (even if some are easy to understand). So I will do a general review about the first part, the generators refactor in #4087 since it could be merged first. And let this issue for the launcher refactor part (since it seems to require the other changes, I hope you are ok with this).

**Comment by akedin** [comment]
@iemejia @kennknowles , I have rebased and created a separate PR for the largest commit here: https://github.com/apache/beam/pull/4165 . Would this be easier to review? If yes, then I'll split this PR into few more smaller ones and close this one. 

@xumingming I addressed the feedback in https://github.com/apache/beam/pull/4165

**Comment by kennknowles** [comment]
I'm afraid I've lost touch with this PR. Would you like to rebase to resolve conflicts and continue to add what is left, or have the changes been merged in other PRs?

**Comment by akedin** [comment]
Closing this PR for now
